### ISSUE TITLE
GPyTorch should handle batched predictions for independent GPs

### ISSUE THREAD

**Issue by jacobrgardner** [cornellius-gp/gpytorch#134]
I'm spinning off a separate issue from the discussion in #130 . GPyTorch should handle `k x n x d`  training data, `k x n` labels, and `k x t x d` test points **at test time** as though we have `k` separate Gaussian processes, each with a training set and a set of test points. 

I think handling this at training time would be much more challenging since we'd need to call backward on a list of `k` losses, but the use case @Balandat  mentioned has fixed hyperparameters, so we only need test time support.

I am marking this as a bug, because I believe @samuelstanton has a fix for this, although it may involve some for loops so we may want to fix it a different way. I think Sam mentioned he actually had a batched solution?

**Comment by Balandat** [comment]
Ideally, we could also to do this at train time, b/c this would gives you a way to do true cross-validation. In this case, if you don't reuse the same underlying tensor, but construct one that is copied from the respective folds of the training data for each of the batches, would it not be possible to just ball backward on the sum of the losses since the gradients will be independent by construction?

**Comment by jacobrgardner** [comment]
We'd have to be careful to replicate all of our `Parameters` if we wanted to do batch training like that. It's possible, but I'll add the `enhancement` label, since handling things at training time will likely require some extra work.

**Comment by Balandat** [comment]
So I was looking into this, and it appears to make this work there are some issues with how means are being treated compared to the covariances. In particular, while `LazyVariable`'s `exact_predictive_covar` properly distinguishes batch mode from non-batch mode and does the subsetting of the full kernel matrix batch-aware, this does not seem to be the case for `exact_predictive_mean`.

Also, should we think about changing how the means are represented internally? If I'm correct it looks like currently everything is geared towards having a prior on the means that is independent of the parameters. While there is no immediate need to implement non-constant mean priors, there are some use cases (see e.g. [1]), and it would be nice to make changes accommodating this earlier rather than later. 

[1] *Unbounded bayesian optimization via regularization. In A. Gretton and C. C. Robert, editors, Proceedings of the 19th International Conference on Artificial Intelligence and Statistics, volume 51 of Proceedings of Machine Learning Research, pages 1168–1176, Cadiz, Spain, 09–11 May 2016. PMLR. URL http://proceedings.mlr.press/v51/shahriari16.html*

**Comment by jacobrgardner** [comment]
I'll take this issue next after I take a look at the priors PR. Getting the predictive means to work in batch should be easy enough.

Can you elaborate on where you think the means are not implemented correctly for non constant means / Do you have specific lines? As far as I can tell, you should be able to implement whatever mean module you need that does whatever you want in forward, non constant or otherwise.

**Comment by Balandat** [comment]
Hmm, I guess I am conflating the batch prediction with the non-constant mean issue. The basic mean api will indeed support non-constant mean functions.
What I am concerned by is how `n_train` is inferred as `n_train = train_labels.size(0)` here, in batch mode that seems like it would be doing the wrong thing: https://github.com/cornellius-gp/gpytorch/blob/7f62373676126373fff178c3f3d8b3db8561cc88/gpytorch/lazy/lazy_variable.py#L291-L298


At a higher level, how do you imagine the flow for using batch predictions to look like? Train a model in non-batch mode, grab the optimized hyperparams, instantiate a new model with the batched train_x/train_y, initialized the hyperparameters, and then predict?

**Comment by jacobrgardner** [comment]
Definitely agree that exact_predictive_mean is a problem for supporting batch predictions at the moment. Shouldn't be too bad for me to fix it up though, hopefully!

As for interface, you recently added a `set_train_data` method to GP models in a PR, so I imagine you could just do a call to that, and then predict.

I'll spend a bit of time thinking about the training side of things. The main challenge there should be an interface once -- all of the heavy lifting internals (like our actual GP math stuff) definitely operate in batch mode, so it should be just a matter of finding an interface we like that changes the batch size on all the hyperparameters (plus probably some minor fixes here and there like with the predictive mean, of course).

**Comment by Balandat** [comment]
Yeah, currently `set_train_data` has some checks that would prevent it from changing the dimensions of the data, but I can relax these checks / maybe retain a "strict" mode. 

W.r.t. the interface - I was facing a similar issue with the dimension of the buffers for the priors, and I initially had a `resize` method that would automatically redefine parameters appropriately given a size. While this ended up being a little overkill and too much magic for that use case, something like this might be useful for the multi-batch training.
